tools/deleteorphanlogcommand.class.php

```diff
@@ -0,0 +1,172 @@
+         // Ask for confirmation (unless --no-interaction)
+         $output->writeln([
+               __('You are about to delete orphan logs of GLPI log table (glpi_logs).'),
+               __('The first step check if orphans exists in database.'),
+               __('The final step is the definitive deletion of orphans (if -f or --force arguments are used).')
+            ]
+         );
+
+         /**
+          * @var QuestionHelper $question_helper
+          */
+         $question_helper = $this->getHelper('question');
+         $run = $question_helper->ask(
+            $input,
+            $output,
+            new ConfirmationQuestion(
+               '<comment>' . __('Do you want to launch operation ?') . ' [yes/No]</comment>',
+               false
+            )
+         );
+         if (!$run) {
+            $output->writeln(
+               '<comment>' . __('Operation aborted.') . '</comment>',
+               OutputInterface::VERBOSITY_VERBOSE
+            );
+            return 0;
+         }
+      }
+
+      return !$this->deleteOrphan($input, $output);
+   }
+
+   private function deleteOrphan(InputInterface $input, OutputInterface $output): bool {
+
+      $globalCount = 0;
+      $listTable = $this->db->listTables();
+      $force = $this->input->getOption('force');
+
+      if (!$force) {
+         $output->writeln(
+            '<comment>' . __('Effective deletion disabled (add -f or --force)') . '</comment>'
+         );
+      }
+
+      $output->writeln(
+         '<comment>' . __('Check if orphans exists.') . '</comment>'
+      );
+
+      while ($table = $listTable->next()) {
+         $tablename = $table['TABLE_NAME'];
+         $itemtype = getItemTypeForTable($tablename);
+
+         $result = $this->db->query(sprintf("select id from glpi_logs as log where log.items_id not in (SELECT id FROM %s) and log.itemtype = '%s';", $tablename, $itemtype));
+         $globalCount += $this->db->numrows($result);
+
+         if ($this->db->numrows($result) > 0) {
+            $output->writeln(
+               sprintf(
+                  '<info>' . __('Found %s record(s) for itemtype %s') . '</info>',
+                  $this->db->numrows($result),
+                  $itemtype
+               )
+            );
+
+            if ($force) {
+
+               $progress_bar = new ProgressBar($output);
+               foreach ($progress_bar->iterate($result) as $log) {
+
+                  $this->db->query(sprintf("DELETE FROM glpi_logs where id = %s ;", $log['id']));
+                  $this->writelnOutputWithProgressBar(
+                     sprintf(
+                        __('Deleting orphan with ID %s for itemtype "%s"'),
+                        $log['id'], $itemtype
+                     ),
+                     $progress_bar,
+                     OutputInterface::VERBOSITY_VERBOSE
+                  );
+               }
+
+               $progress_bar->finish();
+               $this->output->write(PHP_EOL);
```

Comment:
Instead of deleting one by one, maybe you could do something like
```php
$this->db->delete(
    'glpi_logs',
   [
      ['NOT' => ['items_id' => new QuerySubQuery(['SELECT' => 'id', 'FROM' => $tablename])]],
      'itemtype' => itemtype,
   ]
);
```

Reply:
I wan't to keep this (line per line), to get ProgressBar  and to know the state of the deletion

Reply:
Indeed, the progress bar may be usefull if there are a huge amount of row to delete, but if you have thousand orphaned logs lines, you will do too many requests.

Maybe you can delete items 100 by 100 (or 1000 by 1000 if query is not too long).

```php
$progress_bar = new ProgressBar($output);

$ids = [];
do {
   foreach ($result as $row) {
      $ids[] = $log['id'];
      if (count($ids) === 100) {
         break;
      }
   }
   $this->db->->delete('glpi_logs', ['id' => $ids]);
   $progress_bar->advance(100);
} while (count($ids) > 0);

$progress_bar->finish();
```

Reply:
done